### PR TITLE
[misc] Layerwise profile updates

### PR DESCRIPTION
**Pull request by varun-sundar-rabindranath**
Layerwise profile - changes and updates:
 1. Add ability to profile the scenario where the engine is processing requests of different output-lengths.
     -  Remove `--output-length` from offline_profile.py CLI args. 
     - Add sub-commands `run_num_steps` and `run_to_completion` in its place. 
     - `run_num_steps` captures the user-intent more clearly than `--output-length`. i.e. Profile `n` engine steps. 
     -  `run_to_completion` lets the user specify the number of requests the engine should complete every decode step. This provides layer-wise profile information for a range of batch-sizes.
 2. Split gemm_ops  into cutlass-gemm-ops and gemm-ops
 3. Fix ProfileContext read in visualization code.
 4. Bug fix : Incorrect use of rstrip()
 5. Update layerwise_profile and visualization code to store and display metadata.

Examples:
`run_num_steps` sub-command:
```
model=neuralmagic/Meta-Llama-3-8B-Instruct-FP8
python3 examples/offline_profile.py --model ${model} --batch-size 128 --prompt-len 64 --json llama-8b-fp8-num-steps --csv llama-8b-fp8-num-ste    ps --enforce-eager run_num_steps -n 5
```
Graph:
command : 
```
python3 tools/profiler/visualize_layerwise_profile.py --json-trace llama-8b-fp8-num-steps.json --output-directory profile-breakdown-num-steps     --level kernel --plot-metric cuda_time_ms --step-plot-interval 1
``` 
prefill
decode_steps

`run_to_completion` subcommand:
```
model=neuralmagic/Meta-Llama-3-8B-Instruct-FP8
python3 examples/offline_profile.py --model ${model} --batch-size 128 --prompt-len 64 --json llama-8b-fp8-to-completion --csv llama-8b-fp8-to-    completion --enforce-eager run_to_completion -n 16 
```
Graphs
```
python3 tools/profiler/visualize_layerwise_profile.py --json-trace llama-8b-fp8-to-completion.json --output-directory profile-breakdown-to-com    pletion --level kernel --plot-metric cuda_time_ms --step-plot-interval 1
```
prefill
decode_steps

### CODE REVIEW

**Comment by varun-sundar-rabindranath** [comment]
@LucasWilkinson  i made some updates to layerwise-profile . Can you please take a look. Thanks !

File: examples/offline_profile.py [hunk 1]
```diff
@@ -261,12 +269,16 @@ def abort_requests():
...
+
+    parser.add_argument("--max-output-len",
+                        type=int,
+                        default=OUTPUT_LEN_DEFAULT,
+                        help="Maximum output length of the requests")
+
+    parser.add_argument("--min-output-len",
+                        type=int,
+                        default=OUTPUT_LEN_DEFAULT,
+                        help="Minimum output length of the requests")
```
Comment (LucasWilkinson) [inline examples/offline_profile.py:281]: could we maybe keep a `--output-len` option? thats mutually exclusive  with `--max-output-len` and `--min-output-len`?, seems a bit cumbersome if I want say all to have output-len of 8 to do `--max-output-len 8 --min-output-len 8`

File: tools/profiler/visualize_layerwise_profile.py [hunk 2]
```diff
@@ -151,16 +151,18 @@ def is_quant(op_name: str):
            "scaled_int8_quant" in op_name:
             return True
 
+    def is_cutlass_gemm_op(op_name: str):
+        return "void cutlass::Kernel" in op_name or \
+           "void cutlass::device_kernel" in op_name
```
Comment (LucasWilkinson) [inline tools/profiler/visualize_layerwise_profile.py:156]: maybe we should check for `gemm` in the name too? not that we use cutlass for anything else right now, but might prevent future confusion if we use cutlass for convolution or something

**Comment by LucasWilkinson** [comment]
Left a few comments otherwise LGTM. Thanks, these seem like nice improvements!

**Review by LucasWilkinson — APPROVED** [review]

File: examples/offline_profile.py (hunk 1 above)
Comment (varun-sundar-rabindranath) [reply examples/offline_profile.py]: Hey. I have updated to CLI args to pass in `num_steps` directly. I believe it is better in capturing the intent. PTAL.

File: tools/profiler/visualize_layerwise_profile.py (hunk 2 above)
Comment (varun-sundar-rabindranath) [reply tools/profiler/visualize_layerwise_profile.py:156]: Yeah, you are right. Ill add it.

**Comment by varun-sundar-rabindranath** [comment]
Hi @LucasWilkinson  - I've made some non trivial changes. PTAL, thanks !

File: examples/offline_profile.py [hunk 3]
```diff
@@ -34,23 +40,155 @@ def get_dtype(dtype: str):
...
+OutputLen_NumReqs_Map: TypeAlias = Dict[int, int]
+def compute_request_output_lengths(batch_size: int, step_requests: List[int]) \
+      -> OutputLen_NumReqs_Map:
+    """
+    Given the number of requests, batch_size, and the number of requests
+    that each engine-step should process, step_requests, determine the
+    output lengths of the requests such that step_request is honoured.
+
+    Example: 
+    if batch size = 32 and step_request = [128, 128, 96, 64, 32, 1]
```
Comment (LucasWilkinson) [inline examples/offline_profile.py:52]: I'm confused by the 'batch size = 32', when step_request[0] = 128, this doesn't seem to align with the '--complete-num-requests-per-step' comment where it is batch size 128

Comment (varun-sundar-rabindranath) [reply examples/offline_profile.py]: My mistake. It should be "batch_size = 128" ... fixed now 👍

**Review by LucasWilkinson — APPROVED** [review]
LGTM, thanks for the improvements!

**Review by robertgshaw2-redhat — APPROVED** [review]

File: examples/offline_profile.py [hunk 4]
```diff
@@ -187,10 +339,10 @@ def abort_requests():
             for idx, dr in enumerate(decode_results_list):
                 json_dict[f"decode_{idx + 1}"] = dr.convert_stats_to_dict()
 
-        for idx, dr in enumerate(decode_results_list[1:]):
-            json_dict[f"decode_{idx + 1}"] = dr.convert_stats_to_dict()
-
-        with open(json_output.rstrip(".json") + ".json", "w+") as f:
+        # Add .json to json_output filename if it doesn't exist already.
+        json_output_file = json_output if json_output.endswith(
+            '.json') else json_output + '.json'
```
Comment (mgoin) [inline examples/offline_profile.py:344]: nit: do we print out the name of the saved output file?

**Review by mgoin — APPROVED** [review]
LGTM for merge! Nits can be followed in the future

**Pull request merged**